py/query.py

```diff
@@ -0,0 +1,265 @@
+
+      cursor.execute("select exists(select name from pg_prepared_statements where name = 'query_by_range');")
+
+      if cursor.fetchone()[0] == False:
+        try:
+          prepare_statement = "PREPARE query_by_range AS SELECT segment_id, prev_segment_id, start_time, " \
+                              "start_time_dow, start_time_hour, end_time, speed, length FROM segments where " \
+                              "start_time >= $1 and start_time < $2 order by segment_id, prev_segment_id;"
+          cursor.execute(prepare_statement)
+          sql_conn.commit()
+        except Exception as e:
+          raise Exception("Can't create prepare statements")
+    except Exception as e:
+      raise Exception("Can't check for prepare statements")
+    self.sql_conn = sql_conn
+    sys.stdout.write("Created prepare statements.\n")
+    sys.stdout.flush()
+
+  def process_request_thread(self):
+    self.make_thread_locals()
+    while True:
+      request, client_address = self.requests.get()
+      ThreadingMixIn.process_request_thread(self, request, client_address)
+
+  def handle_request(self):
+    try:
+      request, client_address = self.get_request()
+    except socket.error:
+      return
+    if self.verify_request(request, client_address):
+      self.requests.put((request, client_address))
+
+#enable threaded server
+class ThreadedHTTPServer(ThreadPoolMixIn, HTTPServer):
+  pass
+
+#custom handler for getting routes
+class StoreHandler(BaseHTTPRequestHandler):
+
+  #boiler plate parsing
+  def parse_segments(self, post):
+    #split the query from the path
+    try:
+      split = urlparse.urlsplit(self.path)
+    except:
+      raise Exception('Try a url that looks like /action?query_string')
+    #path has the action in it
+    try:
+      if split.path.split('/')[-1] not in actions:
+        raise
+    except:
+      raise Exception('Try a valid action: ' + str([k for k in actions]))
+    #handle POST
+    if post:
+      body = self.rfile.read(int(self.headers['Content-Length'])).decode('utf-8')
+      return json.loads(body)
+    #handle GET
+    else:
+      params = urlparse.parse_qs(split.query)
+      if 'json' in params:      
+        return json.loads(params['json'][0])
+    raise Exception('No json provided')
+
+  #parse the request because we dont get this for free!
+  def handle_request(self, post):
+    #get the query data
+    segments = self.parse_segments(post)
+
+    try:   
+      # get the provider. 
+      list_of_ids = segments['segments'] if 'segments' in segments else None
+      start_date_time = segments.get('start_date_time', None)
+      end_date_time = segments.get('end_date_time', None)
+      cursor = self.server.sql_conn.cursor()
+
+      # send it to the cursor.
+
+      if None in (start_date_time, end_date_time):
+        cursor.execute("execute query_by_id (%s)",(list_of_ids,))
+      else:
+        s_date_time = calendar.timegm(time.strptime(start_date_time,"%Y-%m-%d"))
```

Comment:
This will all be stored and queried in UTC, right?

If so, it's probably best to support queries that go down to at least hours. That way a client has the option to query for start/end that matches day boundaries in local timezones.